Preserve nanosecond resolution when encoding/decoding times

kmuehlbauer opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes #7817, 
- [x] closes #7942, 
- [x] closes #7790, 
- [x] closes #6191
- [x] closes #7096
- [x] closes #1064
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`


Closes #7098 as not needed, but worked around to preserve the fast int-based timedelta calculation.

kmuehlbauer commented:
@spencerkclark I'd appreciate if you could have a look here. All but one test pass, but I can't immediately see what that test is doing. Looks like mismatched dtypes on the attributes. If you have any suggestions how to possibly improve, please let me know. I've not added tests here, yet.

kmuehlbauer commented:
I've reset the order of coders to the initial behaviour. Instead the times are special cased in the CFMaskCoder. Locally it works, but I'll only trust the CI.

kmuehlbauer commented:
All tests have passed. Rebased now on latest main. The issue described in #7817 is resolved. Ready for first reviews.

spencerkclark commented:
Thanks @kmuehlbauer -- I just wanted to give you a heads up that I'm pretty busy this week.  Hopefully I'll get a free moment to look at this more closely next week.

kmuehlbauer commented on xarray/coding/variables.py:
```diff
@@ -235,10 +235,19 @@ def encode(self, variable: Variable, name: T_Name = None):
+        if isinstance(units, str) and "since" in units:
+            encoded_dtype = encoding.pop("dtype")
+            if encoded_dtype is not None and encoded_dtype != data.dtype:
+                data = np.asarray(data, dtype=encoded_dtype)
```
Ah, I'll see how to make this lazy. Thanks.

kmuehlbauer replied:
@Illviljan I've checked the other encoders:

- `CFDatetimeCoder`/`CFTimedeltaCoder` using `np.asarray` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/times.py#L636
- `CFScaleOffsetCoder` uses `data.astype` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/variables.py#L333
- `UnsignedIntegerCoder` uses `duck_array_ops.astype` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/variables.py#L380
- `NonStringCoder` uses `data.astype` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/variables.py#L510
- `BooleanCoder` uses `duck_array_ops.astype` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/variables.py#L452

Now I'm a bit confused what to choose, maybe simple `.astype` is good enough? As `CFDatetimeCoder` is first run and already uses `np.asarray` this might work as is. Anyway to me it looks like those different casts might be aligned or they all serve a different purpose.

Illviljan replied:
But the function has a Variable as an input, and that can be any duckarray (sparse, dask, etc)

I think you should go with `duck_array_ops.astype`. array api standard seems to only have `xp.astype`, https://data-apis.org/array-api/latest/API_specification/generated/array_api.astype.html.

kmuehlbauer replied:
Thanks @Illviljan, much appreciated.

kmuehlbauer commented:
Thanks for the heads-up, @spencerkclark. No worries, I need to apply some changes anyway as it turns out.

dcherian commented on xarray/coding/variables.py:
```diff
@@ -273,9 +282,19 @@ def decode(self, variable: Variable, name: T_Name = None):
-
-            dtype, decoded_fill_value = dtypes.maybe_promote(data.dtype)
-
+            units = attrs.get("units", None)
```
Can we move `_FillValue` to `encoding` in

https://github.com/pydata/xarray/blob/51554f2638bc9e4a527492136fe6f54584ffa75d/xarray/coding/times.py#L723

 and then `CFMaskCoder` would not do anything?

kmuehlbauer replied:
That should work, but we might have to do the casting to final dtype (which is done in `NonStringCoder`) in `CFDatetimeCoder` too. 

I'll have a go on this suggestion, thanks @dcherian.

dcherian replied:
but at decode, the final dtype is always `datetime64[ns]`?

kmuehlbauer replied:
In most cases, but it can also be an array of cftime-objects. I've once again restructured, but I'm not sure if this fit's all cases.

kmuehlbauer commented:
@dcherian You were right from the beginning, changing order for decoding and handling `_FillValue` in `CFDatetimeCoder` seems to be one working solution with minimal code changes.

If the CI is happy I'll add tests to cover for the nanosecond issues in #7817.

kmuehlbauer commented on xarray/coding/times.py:
```diff
@@ -669,8 +663,10 @@ def encode_cf_datetime(
+        # to preserve integer dtype if possible (GH 4045)
+        # NaT prevents us from using datetime64 directly, but we can safely coerce
+        # to int64 in presence of NaT, so we just dropna before check.
+        if np.all(time_deltas.dropna() % time_delta == np.timedelta64(0, "ns")):
```
Note to self, .dropna() removes any NaT from time_deltas. As consequence if-clause returns True if int64 can be preserved. Unfortunately the floor division still returns float64 (as we have NaT inside). We would need to do something like:

```python
num = time_deltas.astype(np.int64) // time_delta.astype(np.int64)
```
to keep int64. IMHO we can do this as `NaT` translates to `-9223372036854775808` (minimal int64) and back. It might be possible to special case NaT to use it as FillValue.

kmuehlbauer replied:
This check has been removed. It's tackled by ref-dat and unit comparison.

spencerkclark reviewed (commented):
Thanks @kmuehlbauer -- I'm catching up here a little bit.  I think if we want to swap the order of decoding datetimes / timedeltas with decoding the mask information, we will also need to decode the `_FillValue` into something that has a datetime or timedelta type.  We may not have great test coverage in this area.  

As a concrete example I'm thinking about something like the following:
```
>>> import numpy as np; import xarray as xr
>>> times = [np.datetime64("2000-01-01"), np.datetime64("NaT")]
>>> da = xr.DataArray(times, dims=["time"], name="foo")
>>> da.encoding["dtype"] = np.float64
>>> da.encoding["_FillValue"] = 20.0
>>> da.to_dataset().to_netcdf("test-encode.nc")
```

On the current main branch we get the following behavior:
```
>>> xr.open_dataset("test-encode.nc").foo.values
array(['2000-01-01T00:00:00.000000000',                           'NaT'],
      dtype='datetime64[ns]')
>>> xr.open_dataset("test-encode.nc", decode_cf=False).foo.values
array([ 0., 20.])
```

With this branch we get:
```
>>> xr.open_dataset("test-encode.nc").foo.values
array(['2000-01-01T00:00:00.000000000',                           'NaT'],
      dtype='datetime64[ns]')
>>> xr.open_dataset("test-encode.nc", decode_cf=False).foo.values
array([ 0., nan])
```

Note the difference in the case where `decode_cf=False`.  It seems like with this branch `NaN` is written out to disk when the fill value prescribes `20.0`, which is incorrect.  I think maybe the fact that `NaN` is being written out to disk is hiding the issue that the `_FillValue` needs to be decoded?

(There is probably a way to test this without doing I/O, but this was the quickest example that came to mind).

kmuehlbauer commented:
Thanks @spencerkclark for taking the time. NaN has been written to disk (as you assumed). Let's have another try next week.

kmuehlbauer commented:
@spencerkclark With current master I get the following `RuntimeWarning` running your code example:

- on encoding (calling `to_netcdf()`):

```python
/home/kai/miniconda/envs/xarray_311/lib/python3.11/site-packages/xarray/coding/times.py:618: RuntimeWarning: invalid value encountered in cast
  int_num = np.asarray(num, dtype=np.int64)
```

- on decoding (calling `open_dataset()`):

```python
/home/kai/miniconda/envs/xarray_311/lib/python3.11/site-packages/xarray/coding/times.py:254: RuntimeWarning: invalid value encountered in cast
  flat_num_dates_ns_int = (flat_num_dates * _NS_PER_TIME_DELTA[delta]).astype(
/home/kai/miniconda/envs/xarray_311/lib/python3.11/site-packages/xarray/coding/times.py:254: RuntimeWarning: invalid value encountered in cast
  flat_num_dates_ns_int = (flat_num_dates * _NS_PER_TIME_DELTA[delta]).astype(
```

The latter was discussed in #7098 (casting float64 to int64), the former was aimed to be resolved with this PR. 

I'll try to create a test case using `Variable` and the respective encoding/decoding functions without involving IO (per your suggestion @spencerkclark).

kmuehlbauer commented:
The example below is only based on Variable and the cf encode/decode variable functions.

```python
import xarray as xr
import numpy as np

# create DataArray
times = [np.datetime64("2000-01-01", "ns"), np.datetime64("NaT")]
da = xr.DataArray(times, dims=["time"], name="foo")
da.encoding["dtype"] = np.float64
da.encoding["_FillValue"] = 20.0

# extract Variable
source_var = da.variable
print("---------- source_var ------------------")
print(source_var)
print(source_var.encoding)

# encode Variable
encoded_var = xr.conventions.encode_cf_variable(source_var)
print("\n---------- encoded_var ------------------")
print(encoded_var)

# decode Variable
decoded_var = xr.conventions.decode_cf_variable("foo", encoded_var)
print("\n---------- decoded_var ------------------")
print(decoded_var.load())
```

```python
/home/kai/miniconda/envs/xarray_311/lib/python3.11/site-packages/xarray/coding/times.py:618: RuntimeWarning: invalid value encountered in cast
  int_num = np.asarray(num, dtype=np.int64)
/home/kai/miniconda/envs/xarray_311/lib/python3.11/site-packages/xarray/coding/times.py:254: RuntimeWarning: invalid value encountered in cast
  flat_num_dates_ns_int = (flat_num_dates * _NS_PER_TIME_DELTA[delta]).astype(
/home/kai/miniconda/envs/xarray_311/lib/python3.11/site-packages/xarray/coding/times.py:254: RuntimeWarning: invalid value encountered in cast
  flat_num_dates_ns_int = (flat_num_dates * _NS_PER_TIME_DELTA[delta]).astype(

---------- source_var ------------------
<xarray.Variable (time: 2)>
array(['2000-01-01T00:00:00.000000000',                           'NaT'],
      dtype='datetime64[ns]')
{'dtype': <class 'numpy.float64'>, '_FillValue': 20.0}
dtype num float64

---------- encoded_var ------------------
<xarray.Variable (time: 2)>
array([ 0., 20.])
Attributes:
    units:       days since 2000-01-01 00:00:00
    calendar:    proleptic_gregorian
    _FillValue:  20.0

---------- decoded_var ------------------
<xarray.Variable (time: 2)>
array(['2000-01-01T00:00:00.000000000',                           'NaT'],
      dtype='datetime64[ns]')
{'_FillValue': 20.0, 'units': 'days since 2000-01-01 00:00:00', 'calendar': 'proleptic_gregorian', 'dtype': dtype('float64')}
```

spencerkclark commented:
Great, yeah, that's a nice example without writing to disk.  Indeed I saw those warnings too, but omitted them in my earlier message to focus on the encoding issue (sorry about that).  I agree that these are something we should address.

spencerkclark commented:
One other tricky edge case that occurs to me is one where an extreme fill value (e.g. `1e30`) is used for floating point fields.  If we decode the times first, it might appear that the dates cannot be represented as nanosecond-precision values, but in reality they would be.  We may need to think more about how to handle this edge case in addition to #7817.

kmuehlbauer commented:
@spencerkclark I've changed the approach and rebased on current main. 

- the order of decoding and encoding is not changed
- On decoding, `_FillValue` will be replaced with `NaT` (np.iinfo(np.int64)) in CFMaskCoder,
  need to check, if this works for all cases.
- on decoding, NaN/NaT is special handled in CDDatetimeCoder to prevent undefined cast NaN -> int
- on encoding, NaN/NaT is special handled to prevent unnecessary cast to float and preserve nanosecond resolution. 
- on encoding, the user is warned if the times can't be serialized faithfully with the current `units`-string
- on encoding, _FillValue is special handled in CFMaskCoder to fill it into NaN/NaT

I had to place a workaround into `coerce_nc3_dtype` to allow for int64 (with NaT) to resolve to float64 with nan as before.

This should resolve most of the cast-issues and the nanosecond resolution issues.

I'd very much appreciate if you could have a closer look here and suggest which parts of the code need rework/tests. I'll continue next week with adding tests with examples from above and other issues related to this.

spencerkclark commented on xarray/coding/variables.py:
```diff
@@ -236,19 +236,32 @@ def encode(self, variable: Variable, name: T_Name = None):
             fill_value = pop_to(encoding, attrs, "_FillValue", name=name)
             if not pd.isnull(fill_value):
-                data = duck_array_ops.fillna(data, fill_value)
+                if is_date:
```
Do we also need to ensure that `data` has an integer dtype here?
```suggestion
                if is_date and data.dtype.kind in "iu":
```

spencerkclark commented on xarray/coding/variables.py:
```diff
@@ -236,19 +236,32 @@ def encode(self, variable: Variable, name: T_Name = None):
             fill_value = pop_to(encoding, attrs, "missing_value", name=name)
             if not pd.isnull(fill_value) and not fv_exists:
-                data = duck_array_ops.fillna(data, fill_value)
+                if is_date:
```
```suggestion
                if is_date and data.dtype.kind in "iu":
```

spencerkclark reviewed (commented):
Nice, thanks for the update @kmuehlbauer.  I think this is going in the right direction.  I left a few initial comments.

I'm also trying to think about how this fits in with `cftime.datetime` arrays, though I'm realizing that decoding / encoding arrays with missing values does not really work in any capacity currently for those types of arrays, so it's maybe safe to continue to punt on that until Unidata/cftime#145 is addressed.  

I don't think there should be any risk in replacing missing values with `np.iinfo(np.int64).min` in arrays prior to decoding, since cftime relies on `np.timedelta64` internally in `num2date`, and so encountering those values will end up raising anyway (either due to overflow or an encounter with `None`):
```
>>> cftime.num2date([np.iinfo(np.int64).min], units="seconds since 2000-01-01")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "src/cftime/_cftime.pyx", line 616, in cftime._cftime.num2date
  File "src/cftime/_cftime.pyx", line 458, in cftime._cftime.scale_times
OverflowError: time values outside range of 64 bit signed integers

>>> cftime.num2date([np.iinfo(np.int64).min], units="microseconds since 2000-01-01")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "src/cftime/_cftime.pyx", line 630, in cftime._cftime.num2date
  File "src/cftime/_cftime.pyx", line 499, in cftime._cftime.decode_dates_from_array
TypeError: unsupported operand type(s) for +: 'cftime._cftime.DatetimeGregorian' and 'NoneType'
```

which is already what happens if a `np.nan` value is found:

```
>>> cftime.num2date([np.nan], units="seconds since 2000-01-01")
<stdin>:1: RuntimeWarning: invalid value encountered in cast
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "src/cftime/_cftime.pyx", line 630, in cftime._cftime.num2date
  File "src/cftime/_cftime.pyx", line 499, in cftime._cftime.decode_dates_from_array
TypeError: unsupported operand type(s) for +: 'cftime._cftime.DatetimeGregorian' and 'NoneType'
```

Encoding is a little more awkward, since I think it does kind of work currently for the case where one explicitly specifies the units and calendar (otherwise it fails):


```
>>> xr.coding.times.encode_cf_datetime([None, cftime.DatetimeNoLeap(2000, 1, 1)], units="days since 2000-01-01", calendar="noleap")
/Users/spencer/software/xarray/xarray/coding/times.py:618: RuntimeWarning: invalid value encountered in cast
  int_num = np.asarray(num, dtype=np.int64)
(array([nan,  0.]), 'days since 2000-01-01', 'noleap')
```

But I think this is OK since the presence of `np.nan` means this needs to be a floating point array.

kmuehlbauer commented on xarray/coding/times.py:
```diff
@@ -635,32 +655,41 @@ def encode_cf_datetime(
+                    f"Times can't be serialized faithfully with requested units {units!r}. "
+                    f"Resolution of {needed_delta!r} needed. "
+                    f"Serializing timeseries to floating point."
+                )
```
Thanks @spencerkclark, done.

kmuehlbauer commented on xarray/coding/variables.py:
```diff
@@ -275,7 +288,11 @@ def decode(self, variable: Variable, name: T_Name = None):
 
-            dtype, decoded_fill_value = dtypes.maybe_promote(data.dtype)
+            # special case DateTime to properly handle NaT
+            if "since" in str(attrs.get("units", "")):
```
Thanks, done!

kmuehlbauer commented on xarray/backends/netcdf3.py:
```diff
@@ -62,6 +62,13 @@ def coerce_nc3_dtype(arr):
     dtype = str(arr.dtype)
     if dtype in _nc3_dtype_coercions:
         new_dtype = _nc3_dtype_coercions[dtype]
+        # check if this looks like a time with NaT
+        # and transform to float64
+        if np.issubdtype(dtype, np.int64):
+            mask = arr == np.iinfo(np.int64).min
+            if mask.any():
+                arr = np.where(mask, np.nan, arr)
+                return arr
```
@spencerkclark I agree, your argument with the non-time related int64 containing `np.iinfo(np.int64).min` is very valid. OTOH, we do not know how many legacy workflows exist out there which would break. 

In the light of this how about moving this check to the variable-level where we have the attributes available? I'll try to come up with something reasonable.

kmuehlbauer replied:
I've moved the nc3 handling to `encode_nc3_variable`.

kmuehlbauer commented:
@spencerkclark Thanks for the review and your detailed comments. I've applied your suggestions and added some tests. I've added the needed changes for timedelta-coding by accident earlier. So I just fixed that up too.

spencerkclark commented on xarray/coding/times.py:
```diff
@@ -633,32 +654,41 @@ def encode_cf_datetime(
+            needed_delta = _infer_time_units_from_diff(
+                (data_ref_date - ref_date).to_timedelta64()
+            )
+            needed_time_delta = np.timedelta64(
+                1, _netcdf_to_numpy_timeunit(needed_delta)
+            ).astype("timedelta64[ns]")
```
nit: maybe we can write a function (e.g., `_time_units_to_timedelta64`) to handle the conversion from units `str` to `np.timedelta64` value, since this same conversion is also required above.

spencerkclark commented on xarray/coding/times.py:
```diff
@@ -633,32 +654,41 @@ def encode_cf_datetime(
+        # check if times can be represented with given units
+        if data_units != units:
+            data_delta, data_ref_date = _unpack_delta_ref_date(data_units)
+            needed_delta = _infer_time_units_from_diff(
+                (data_ref_date - ref_date).to_timedelta64()
+            )
+            needed_time_delta = np.timedelta64(
+                1, _netcdf_to_numpy_timeunit(needed_delta)
+            ).astype("timedelta64[ns]")
+            if needed_delta != delta and time_delta > needed_time_delta:
+                emit_user_level_warning(
+                    f"Times can't be serialized faithfully with requested units {units!r}. "
+                    f"Resolution of {needed_delta!r} needed. "
+                    f"Serializing timeseries to floating point."
+                )
```
One other thing that occurs to me about this warning is that it will not be triggered for cftime datetimes (an exception will kick us out of this code path beforehand), but perhaps it's better to punt that responsibility upstream to `cftime.date2num` (i.e. no action needed).

kmuehlbauer replied:
Yes, maybe upstream is the right place...

spencerkclark reviewed (commented):
Thanks @kmuehlbauer!  I think I understand / am on board with everything now.  Just a few nits and some suggestions for additional tests.

kmuehlbauer commented on xarray/coding/times.py:
```diff
@@ -573,6 +591,9 @@ def _should_cftime_be_used(
     delta, ref_date = _unpack_netcdf_time_units(units)
+    delta = delta.lower()
+    if not delta.endswith("s"):
+        delta = f"{delta}s"
```
I had to place this check there, since tests would break otherwise. So I think this is covered.

kmuehlbauer commented on xarray/backends/netcdf3.py:
```diff
@@ -88,13 +88,40 @@ def encode_nc3_attrs(attrs):
+    # replaces np.iinfo(np.int64).min with _FillValue or np.nan
+    # this keeps backwards compatibility
+
+    # should we import this from coding.times here?
```
I'm importing now from in coding.variables (where it is currently defined). If there is a better place, please let me know.

kmuehlbauer commented on xarray/backends/netcdf3.py:
```diff
@@ -88,13 +88,40 @@ def encode_nc3_attrs(attrs):
+            if any(tstr in units for tstr in time_strings):
+                mask = data == np.iinfo(np.int64).min
+                if mask.any():
+                    data = np.where(mask, var.attrs.get("_FillValue", np.nan), data)
```
I need to think about this a bit more.

spencerkclark replied:
I guess this existing test may actually take care of it?  Maybe we can even remove the `@arm_xfail` decorator now?  https://github.com/pydata/xarray/blob/8215911a8e1c97e58c3db071276a185b931b4a35/xarray/tests/test_backends.py#L529-L536

kmuehlbauer commented on xarray/tests/test_coding_times.py:
```diff
@@ -1191,3 +1191,59 @@ def test_contains_cftime_lazy() -> None:
+) -> None:
+    # test for GH7817
+    times = [np.datetime64("1970-01-01", "ns"), np.datetime64("NaT"), time]
+    encoding = dict(dtype=dtype, _FillValue=fill_value)
```
I've adapted the current test with additional examples also checking for encoded attrs and decoded encoding. Let me know, if this is good enough.

spencerkclark replied:
Looks good, thanks!

kmuehlbauer commented on xarray/coding/variables.py:
```diff
@@ -275,7 +303,11 @@ def decode(self, variable: Variable, name: T_Name = None):
 
-            dtype, decoded_fill_value = dtypes.maybe_promote(data.dtype)
+            # special case DateTime to properly handle NaT
+            if _is_time_like(attrs.get("units")) and data.dtype.kind in "iu":
```
Added a test for roundtripping this example.

kmuehlbauer commented:
@spencerkclark I've followed up with your suggestion. Not sure if the tests capture everything now, but I think it's looking good so far.

spencerkclark reviewed (approved):
Beautiful @kmuehlbauer -- I appreciate those additional tests, the additional warning, and the further cleanup.  One more suggestion for a minor tweak to a test, and I think this is ready for a what's new entry.  Huge thanks for taking on this gnarly issue!

kmuehlbauer commented:
Thanks all for pushing this to a mergeable state. @spencerkclark :+1: :100:

spencerkclark commented on doc/whats-new.rst:
```diff
@@ -73,6 +77,8 @@ Internal Changes
 
 - Many error messages related to invalid dimensions or coordinates now always show the list of valid dims/coords (:pull:`8079`).
   By `András Gunyhó <https://github.com/mgunyho>`_.
+- Refactor of encoding and decoding times/timedeltas to preserve nanosecond resolution (:pull:`7827`).
```
This was something we tried to accomplish in #4684 -- and I think we were successful for arrays that did not contain missing values (e.g. [this test](https://github.com/pydata/xarray/blob/8215911a8e1c97e58c3db071276a185b931b4a35/xarray/tests/test_coding_times.py#L1045-L1054)) -- but we overlooked this important issue related to arrays that did contain missing values.  We therefore might consider this a bug fix as well:
```suggestion
- Refactor of encoding and decoding times/timedeltas to preserve nanosecond resolution in arrays that contain missing values (:pull:`7827`).
```

spencerkclark commented on doc/whats-new.rst:
```diff
@@ -63,6 +63,10 @@ Bug fixes
 - Fix bug where :py:class:`DataArray` instances on the right-hand side
   of :py:meth:`DataArray.__setitem__` lose dimension names.
   (:issue:`7030`, :pull:`8067`) By `Darsh Ranjan <https://github.com/dranjan>`_.
+- Fixed a bug where casting from ``float`` to ``int64`` (undefined for ``NaN``) lead to varying
```
```suggestion
- Fixed a bug where casting from ``float`` to ``int64`` (undefined for ``NaN``) led to varying
```

dcherian commented:
Amazing work, @kmuehlbauer and @spencerkclark . Thanks!

meteoDaniel commented:
Awesome contribute @kmuehlbauer  :)
